### PR TITLE
Dynamic formats

### PR DESCRIPTION
**Pull request by yuxshao**
Format string specifying punctuation and location of line, function, and filename.

### CODE REVIEW

File: src/logger.h [hunk 1]
```diff
@@ -5,14 +5,13 @@
...
 #include <mutex>
-// TODO: how do you specify defaults? add a logconfig.h to the library
 
 // * Log record : each line of the log
 // * log format : tuple like <Date, Name, Message>
 // * log entity : Date, Name, Message
 #define LOG(severity)                                                          \
-  logger::Logger::getInstance().log<logger::severity>(__func__, __FILE__,      \
-                                                      __LINE__)
+  logger::Logger::getInstance().log<logger::severity>(                         \
```
Comment (andylhxu) [inline src/logger.h:13]: did not use `clang-format` before committing.

File: src/logger.h [hunk 2]
```diff
@@ -26,21 +25,51 @@ enum Severity {
...
-    stream << file << ":" << line << "(" << fn << ")" << ": ";
+  Format(std::ostream &s, InputInfo input_info,
+                  std::mutex &Logging_lock)
+      : stream(s), info(input_info), lock(Logging_lock) {
+      print_prefix(info.file, info.fn, info.line);
   }
   ~Format() { stream << std::endl; }
 
-  template <typename T> Format &operator<<(const T &s) {
+  template <typename S> Format &operator<<(const S &s) {
```
Comment (andylhxu) [inline src/logger.h]: Nice.

**Review by andylhxu — APPROVED** [review]

**Review by andylhxu — APPROVED** [review]

File: src/logger.h [hunk 3]
```diff
@@ -26,21 +25,47 @@ enum Severity {
...
+  int line;
+};
+
+// Format("(%:%:%)", date, line, ip);
+// Format(const char *, f -> string...);
+template <const char *Fmt> class Format {
   std::ostream &stream;
-  const char *fn, *file; int line;
   std::lock_guard<std::mutex> lock;
+  InputInfo info;
```
Comment (yogeshg) [inline src/logger.h]: ContextInfo -- contains the contextual information of the program "where" the log was created.

File: src/logger.h (hunk 1 above)
Comment (yuxshao) [reply src/logger.h:13]: done

**Pull request merged**